mne/time_frequency/psd.py

```diff
@@ -224,12 +225,30 @@ def psd_array_welch(
         window=window,
         mode=mode,
     )
-    x_splits = [arr for arr in np.array_split(x, n_jobs) if arr.size != 0]
+    if np.any(np.isnan(x)):
+        good_mask = ~np.isnan(x)
+        # NaNs originate from annot, so must match for all channels. Note that we CANNOT
+        # use np.testing.assert_allclose() here; it is strict about shapes/broadcasting
+        assert np.allclose(good_mask, good_mask[[0]], equal_nan=True)
+        t_onsets, t_offsets = _mask_to_onsets_offsets(good_mask[0])
+        x_splits = [x[..., t_ons:t_off] for t_ons, t_off in zip(t_onsets, t_offsets)]
+        weights = [
+            split.shape[-1] for split in x_splits if split.shape[-1] >= n_per_seg
+        ]
```

Comment:
was actually just wondering if this is in fact the most accurate weighting.  Perhaps this is better?

```py
        # weights reflect the number of welch windows in each span
        stepsize = n_per_seg - n_overlap
        weights = [(span.shape[-1] - n_overlap) // stepsize for span in x_splits]
        weights = np.array(weights)[np.array(weights) > 0]
```

Old code weights based on total samples in the span (as long as span > n_per_seg). New code weights based on number of welch windows that fit into the span, i.e., should reflect how much data in the span is actually used in the computation.  If that seems correct, I can open a follow-up PR.  cc @agramfort @larsoner

Reply:
The new way seems more correct to me